snorkel/classification/loss.py

```diff
@@ -0,0 +1,122 @@
+    from a 1D tensor of ints to a 2D tensor of probabilities.
+
+    Parameters
+    ----------
+    input
+        A [num_points, num_classes] tensor of logits
+    target
+        A [num_points, num_classes] tensor of probabilistic target labels
+    weight
+        An optional [num_classes] array of weights to multiply the loss by per class
+    reduction
+        One of "none", "mean", "sum", indicating whether to return one loss per data
+        point, the mean loss, or the sum of losses
+
+    Returns
+    -------
+    torch.Tensor
+        The calculated loss
+
+    Raises
+    ------
+    ValueError
+        If an invalid reduction keyword is submitted
+    """
+    num_points, num_classes = input.shape
+    # Note that t.new_zeros, t.new_full put tensor on same device as t
+    cum_losses = input.new_zeros(num_points)
+    for y in range(num_classes):
+        target_temp = input.new_full((num_points,), y, dtype=torch.long)
+        y_loss = F.cross_entropy(input, target_temp, reduction="none")
+        if weight is not None:
+            y_loss = y_loss * weight[y]
+        cum_losses += target[:, y].float() * y_loss
+
+    if reduction == "none":
+        return cum_losses
+    elif reduction == "mean":
+        return cum_losses.mean()
+    elif reduction == "sum":
+        return cum_losses.sum()
+    else:
+        raise ValueError("Keyword 'reduction' must be one of ['none', 'mean', 'sum']")
+
+
+def ce_soft_loss_from_outputs(
+    op_name: str, outputs: Outputs, Y: torch.FloatTensor, active: torch.IntTensor
+) -> torch.Tensor:
+    """Calculate mean cross-entropy loss for active outputs using probabilistic labels
+
+    Parameters
+    ----------
+    op_name
+        The name of the operation whose output should be used for calculating loss
+    outputs
+        The dictionary of operation outputs
+    Y
+        The gold labels (integers) to calculate loss from
+    active
+        A mask of which data points to consider when calculating loss
+
+    Returns
+    -------
+    torch.Tensor
+        The calculated loss
+    """
+    return cross_entropy_soft(outputs[op_name][0][active], Y[active])
+
+
+def cross_entropy_from_outputs(
+    op_name: str, outputs: Outputs, Y: torch.Tensor, active: torch.IntTensor
+) -> torch.Tensor:
+    """Calculate mean cross-entropy loss for the active outputs of the specified op.
+
+    Note: TBD
+    ...gold labels must be integers. To use "soft" probabilistic labels, use cross_entropy_soft() instead.
+
+    Parameters
+    ----------
+    op_name
+        The name of the operation whose output should be used for calculating loss
+    outputs
+        The dictionary of operation outputs
+    Y
+        The gold labels (integers) to calculate loss from
+    active
+        A mask of which data points to consider when calculating loss
+
+    Returns
+    -------
+    torch.Tensor
+        The calculated loss
+    """
+    return F.cross_entropy(outputs[op_name][0][active], (Y.view(-1))[active])
```

Comment:
I've wanted to get rid of the 'active' argument as well. The issue is that doing so puts us in charge of any reduction that happens. e.g., we'll have to decide whether to average or sum the losses in a batch instead of leaving it up to the user, and some loss functions (e.g., ranking loss) are at the batch level rather than the individual example level anyway. Passing `active` is the most flexible (the user can use it or not; all we care about is that their function returns a number at the end of the day), but being more prescriptive (doing the filtering on the model side before the loss function and forcing that we always do the 'mean' reduction) is cleaner. I guess I'm leaning more toward the latter, since it covers the most common use case and makes it easier for the user to swap in their own function for ours (ones less arg to worry about in the signature). @vincentschen any thoughts here?

Reply:
Ah, I remembered the other issue: we pass the function the whole outputs dictionary (so that it can use multiple outputs as in multiple-choice settings or an output from a layer other than the last one as in hierarchical MTL settings ); we don't know beforehand which output(s) to filter and filtering all outputs feels clumsy/wasteful. If we were to make the assumption that we will always calculate losses and predictions using only the output from the last operation, that would simplify a number of things. And come to think of it, I believe that the unusual cases mentioned above (multiple choice, hierarchical MTL) could be handled with a sort of trivial final operation that just identity maps from whatever layers are necessary to the fields the loss function needs. Let's do that in a separate PR, since it has more to do with model execution rather than the probabilistic cross entropy loss.

Reply:
Those are too long to read. Assuming you're on top of the right call here.